Homeworks/Homework_3/script.py

```diff
@@ -0,0 +1,327 @@
+
+    :param csv_path: str, path to the folder where the csv file is located
+    :param csv_data: list, data which has to be written to the csv file
+    """
+    with open(csv_path, 'w', newline='', encoding='utf-8') as csv_file:
+        csv_writer = csv.DictWriter(csv_file, fieldnames=csv_data[0].keys())
+        csv_writer.writeheader()
+        csv_writer.writerows(csv_data)
+
+
+def create_folder_if_not_exist(destination_folder_path):
+    """
+    Create a folder at the given destination path if it does not exist already.
+
+    :param destination_folder_path: str, the path where the folder needs to be created.
+    """
+    if not os.path.exists(destination_folder_path):
+        os.makedirs(destination_folder_path)
+        logging.info(f'Directory {destination_folder_path} was successfully created')
+
+
+def rearrange_data(old_csv_data):
+    """
+    Rearrange data into format 'decades : {countries : [users_data]}'
+
+    :param old_csv_data: list, data from the csv file in the old format
+    :return: dict, data which was rearranged from list format to the dict format
+    """
+    new_data = {}
+    for row in old_csv_data:
+        decade = f'{row["dob.date"][-2]}0-th'
+        country = row['location.country']
+        new_data.setdefault(decade, {})
+        new_data[decade].setdefault(country, [])
+        new_data[decade][country].append(row)
+    return new_data
+
+
+def create_file_name(rearranged_data, folder_decade, folder_country):
+    """
+    Create a specific name for the CSV file based on user's maximum age, average number of years of registration,
+    and the most common id.name.
+
+    :param rearranged_data: list, containing all data about users, organized by decade and country.
+    :param folder_decade: str, the name of the folder representing the decade.
+    :param folder_country: str, the name of the folder representing the country.
+    :return: str, the name of the CSV file based on the provided data.
+    """
+
+    users_data = [(int(user['dob.age']), int(user['registered.age']), user['id.name']) for user in
+                  rearranged_data[folder_decade][folder_country]]
+
+    return f'{max(users_data, key=lambda user_info: user_info[0])[0]}_' \
+           f'{sum([element[1] for element in users_data]) / len(users_data)}_' \
+           f'{Counter([element[2] for element in users_data]).most_common(1)[0][0]}'
+
+
+def create_sub_folders(rearranged_data):
+    """
+    Create folders with csv files based on the new data format
+
+    :param rearranged_data: dict, data on the structure of which
+                            folders are created and csv files into folders are filled
+    """
+    for decade in rearranged_data:
+        decade_path = os.path.join(os.getcwd(), decade)
+        os.makedirs(decade_path)
+        for country in rearranged_data[decade]:
+            country_path = os.path.join(decade_path, country)
+            os.makedirs(country_path)
```

Comment:
you can put to os.path.join as many parts as you like.
I believe the second makedirs can create a nested hierarchy. so, you can delete the first one